rust/lance/src/dataset.rs

```diff
@@ -6553,4 +6553,155 @@ mod tests {
+            schema.clone(),
+            vec![Arc::new(UInt32Array::from_iter_values(0..10))],
+        )
+        .unwrap();
+        let batches = RecordBatchIterator::new(vec![data].into_iter().map(Ok), schema.clone());
+        let mut dataset = Dataset::write(batches, test_uri, None).await.unwrap();
+
+        // Verify we have 1 fragment with id 0
+        assert_eq!(dataset.get_fragments().len(), 1);
+        assert_eq!(dataset.get_fragments()[0].id(), 0);
+        assert_eq!(dataset.manifest.max_fragment_id(), Some(0));
+
+        // Delete all rows
+        dataset.delete("true").await.unwrap();
+
+        // After deletion, dataset should be empty but max_fragment_id preserved
+        assert_eq!(dataset.get_fragments().len(), 0);
+        assert_eq!(dataset.count_rows(None).await.unwrap(), 0);
+        assert_eq!(dataset.manifest.max_fragment_id(), Some(0));
+
+        // Append another fragment
+        let data = RecordBatch::try_new(
+            schema.clone(),
+            vec![Arc::new(UInt32Array::from_iter_values(20..30))],
+        )
+        .unwrap();
+        let batches = RecordBatchIterator::new(vec![data].into_iter().map(Ok), schema.clone());
+        let write_params = WriteParams {
+            mode: WriteMode::Append,
+            ..Default::default()
+        };
+        let dataset = Dataset::write(batches, test_uri, Some(write_params))
+            .await
+            .unwrap();
+
+        // Assert new fragment has id 1, not 0
+        assert_eq!(dataset.get_fragments().len(), 1);
+        assert_eq!(dataset.get_fragments()[0].id(), 1);
+        assert_eq!(dataset.manifest.max_fragment_id(), Some(1));
+    }
+
+    #[rstest]
+    #[tokio::test]
+    async fn test_fragment_id_never_reset() {
+        // Test case 2: Fragment id is never reset, even if all rows are deleted
+        // 1. Create dataset with N fragments
+        // 2. Delete all rows
+        // 3. Append more fragments
+        // 4. Assert new fragments have ids >= N
+
+        let test_dir = tempdir().unwrap();
+        let test_uri = test_dir.path().to_str().unwrap();
+
+        let schema = Arc::new(ArrowSchema::new(vec![ArrowField::new(
+            "i",
+            DataType::UInt32,
+            false,
+        )]));
+
+        // Create dataset with 3 fragments (N=3)
+        let data1 = RecordBatch::try_new(
+            schema.clone(),
+            vec![Arc::new(UInt32Array::from_iter_values(0..10))],
+        )
+        .unwrap();
+        let data2 = RecordBatch::try_new(
+            schema.clone(),
+            vec![Arc::new(UInt32Array::from_iter_values(10..20))],
+        )
+        .unwrap();
+        let data3 = RecordBatch::try_new(
+            schema.clone(),
+            vec![Arc::new(UInt32Array::from_iter_values(20..30))],
+        )
+        .unwrap();
+        let batches = RecordBatchIterator::new(
+            vec![data1, data2, data3].into_iter().map(Ok),
+            schema.clone(),
+        );
```

Comment:
You don't need to create 3 separate batches. The `max_rows_per_file` will handle splitting into fragments.
```suggestion
        // Create dataset with 3 fragments (N=3)
        let data = RecordBatch::try_new(
            schema.clone(),
            vec![Arc::new(UInt32Array::from_iter_values(0..30))],
        )
        .unwrap();
        .unwrap();
        let batches = RecordBatchIterator::new(
            vec![Ok(data)],
            schema.clone(),
        );
```